app/ModelFunctions/LDAPFunctions.php

```
@@ -0,0 +1,377 @@
+<?php
+
+namespace App\ModelFunctions;
+
+use App\Exceptions\LDAPException;
+use App\Models\Configs;
+use App\Models\Logs;
+
+class LDAPFunctions
+{
+	/* @var resource $con holds the LDAP connection */
+	protected $con = null;
+
+	/* @var int $bound What type of connection does already exist? */
+	protected $bound = 0; // 0: anonymous, 1: user, 2: superuser
+
+	/* @var array user info queried from LDAP */
+	protected $user_info = [];
+
+	/**
+	 * Log debug message.
+	 *
+	 * @param string $method
+	 * @param string $line
+	 * @param string $text
+	 */
+	protected function _debug(string $method, string $line, string $text)
+	{
+		if ($this->getConf('debug', '0')) {
+			Logs::notice($method, $line, $text);
+		}
+	}
+
+	/**
+	 * Wraps around ldap_search, ldap_list or ldap_read depending on $scope.
+	 *
+	 * @param resource   $link_identifier
+	 * @param string     $base_dn
+	 * @param string     $filter
+	 * @param string     $scope           can be 'base', 'one' or 'sub'
+	 * @param array|null $attributes
+	 * @param int        $attrsonly
+	 * @param int        $sizelimit
+	 *
+	 * @return resource
+	 */
+	protected function _ldapsearch($link_identifier, $base_dn, $filter, $scope = 'sub', $attributes = null,
+						 $attrsonly = 0, $sizelimit = 0)
+	{
+		if (is_null($attributes)) {
+			$attributes = [];
+		}
+
+		if ($scope == 'base') {
+			return @ldap_read(
```

Please don't use the `@`-operator to silence engine errors. The `@`-operator is legacy and evil. For many functions the `@`-operator has already lost its function (e.g. see https://wiki.php.net/rfc/engine_warnings and https://php.watch/versions/8.0/fatal-error-suppression). As PHP is slowly phasing out this operator, its usage is not future proof. It might be that it won't work for ldap-functions is the next PHP release. Moreover, silencing errors makes tracking down errors only more difficult, because eventually the code will fail somewhere else and then it is only more difficult to find the real cause.

Lychee has only recently moved to use proper exceptions nearly everywhere and all remaining occurrences of the `@`-operator have been nuked. Please use exceptions.

This remark is an example and holds for all similar code lines.

---

I removed all @ operators.

---

I put the @-operators in again, because they are required (see above).